### PR TITLE
Handle complex optimization in SparseAdam by treating complex numbers as 2D real numbers

### PR DESCRIPTION
**Pull request by zaxtax**
This commit partially addresses #65711

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
lint and tests
Sounds good otherwise.

File: torch/optim/_functional.py [hunk 1]
```diff
@@ -74,3 +80,9 @@ def make_sparse(values):
         step_size = lr * math.sqrt(bias_correction2) / bias_correction1
 
         param.add_(make_sparse(-step_size * numer.div_(denom)))
+
+        if torch.is_complex(param):
```
Comment (albanD) [inline torch/optim/_functional.py:84]: These are not needed ;)

File: test/test_optim.py [hunk 2]
```diff
@@ -320,12 +320,22 @@ def _test_complex_optimizer(self, optimizer_constructor):
...
+        if sparse:
+            i = torch.LongTensor([[1, 1]])
+            a1 = torch.sparse_coo_tensor(i, a1, torch.Size([2]))
+            a1_real = a1.to_dense().real.to_sparse()
+            a1_imag = a1.to_dense().imag.to_sparse()
+        else:
+            a1_real = a1.real
+            a1_real = a1.imag
+
+        al_real = al_real.clone().detach()
```
Comment (albanD) [inline test/test_optim.py:337]: ai vs a1

**Review by albanD — COMMENTED** [review]
small changes but LGTM otherwise

**Review by albanD — APPROVED** [review]
Change sounds good. Test typo needs to be fixed and this is good to land!

**Comment by rgommers** [comment]
Test errors are relevant: `RuntimeError: SparseAdam does not support dense gradients, please consider Adam instead`

**Comment by rgommers** [comment]
This still has relevant failures, e.g. on the `linux-bionic-py3.7-clang9 / test (dynamo, 1, 2, linux.2xlarge)` job: 
```
   test_sparse_adam (__main__.TestOptim) ... ERROR (1.980s)
    test_sparse_adam errored - num_retries_left: 2
Traceback (most recent call last):
  File "test_optim.py", line 655, in test_sparse_adam
    def test_sparse_adam(self):
  File "test_optim.py", line 655, in test_sparse_adam
    def test_sparse_adam(self):
  File "test_optim.py", line 667, in test_sparse_adam
    self._test_complex_2d(optim.SparseAdam, sparse=True)
  File "test_optim.py", line 331, in _test_complex_2d
    a1_real = a1.real
NotImplementedError: Could not run 'aten::view_as_real' with arguments from the 'SparseCPU' backend. This could be because the operator doesn't exist for this backend, or was omitted during the selective/custom build process (if using custom build). If you are a Facebook employee using PyTorch on mobile, please visit https://fburl.com/ptmfixes for possible resolutions. 'aten::view_as_real' is only available for these backends: [CPU, Meta, BackendSelect, Python, Functionalize, Named, Conjugate, Negative, ZeroTensor, ADInplaceOrView, AutogradOther, AutogradCPU, AutogradCUDA, AutogradHIP, AutogradXLA, AutogradMPS, AutogradIPU, AutogradXPU, AutogradHPU, AutogradVE, AutogradLazy, AutogradMeta, AutogradPrivateUse1, AutogradPrivateUse2, AutogradPrivateUse3, AutogradNestedTensor, Tracer, AutocastCPU, AutocastCUDA, Batched, VmapMode, PythonTLSSnapshot].
```

**Comment by zaxtax** [comment]
I'm closing this PR has I'm blocked by needing `aten::view_as_real` to be supported by the `SparseCPU` backend.

**Pull request closed without merging**